### ISSUE TITLE
Safely handling all reductions in CUDA & OpenMP code generation

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#430]
This is referring to how the Regent compiler generates CUDA & OpenMP code for the following code pattern:

```
__demand(__openmp)
for i in is do
  r[f(i)].fld += expr
end
```

when `f` is not statically known to be injective. Currently this pattern will be broken down into a read of `r[f(i)].fld`, followed by an addition, followed by a write of `r[f(i)].fld`, which can trigger a race condition if two indices `i1` and `i2` that satisfy `f(i1) = f(i2)` happen to be processed at the same time.

For Soleil-X in particular, this pattern occurs when we want to loop over all particles and accumulate a per-particle value onto the cell containing that particle, and this is implemented like so: https://github.com/stanfordhpccenter/soleil-x/blob/dca0d473d2eb6cbf07b39ab4fbfd18e6ed30f2a9/src/soleil.rg#L4050-L4066.

One possible solution is to provide the user with an "atomic reduction" primitive and have them use it whenever they can't guarantee that `f` is injective. A safer option would be to have the compiler always emit atomic operations when it can't statically prove that `f` is injective (and possibly allow the user to override this decision).

An additional complication that we have to deal with in Soleil-X is that we often want to reduce into array-typed fields, e.g. see https://github.com/stanfordhpccenter/soleil-x/blob/dca0d473d2eb6cbf07b39ab4fbfd18e6ed30f2a9/src/soleil.rg#L4110-L4130. This operation is safe to decompose into 3 separate atomic reductions on each of the 3 elements of the tuple, but AFAIK there is no way to express this currently such that the code generator will pick it up. The code generator could be taught to match a pattern like this one and code-generate it in a special way:
```
for k = 0,3 do
  q[f(i)].fld[k] += expr[k]
end
```
or even just know to consider the array index `k` in a statement like `q[f(i)].fld[k] += ...` as part of the l-value.

**Comment by lightsighter** [comment]
I think I'm personally in favor of the user having to specify the atomic primitive in the case where they demand OpenMP or CUDA code generation. If the user is explicit about the parallelism then they should be explicit about where the synchronization goes. On the other hand, if the compiler does automatic parallelization then I think we probably want the compiler to be safe and put in atomics whenever it can't prove something itself.

**Comment by magnatelee** [comment]
I agree that a `__demand` pragma should change only the performance, not the semantics. Here is what I'm going to do: I'll make the compiler emit an atomic reduction when the region access is made with an opaque index expression and add an `__insist` pragma to avoid the atomic operation when they know it is safe. I can at least try to check if the index expression is affine, but that's just an optimization.

**Comment by magnatelee** [comment]
@elliottslaughter and I briefly discussed about this issue and we both think that an assertion rather than a pragma is a right way to convey the "disjointness" information to the compiler.

**Comment by magnatelee** [comment]
@manopapad I pushed an initial fix (d31a9166) that at least covers Soleil-X. The "centeredness" check in this fix is not yet complete, so I'll keep this issue open until I pushed a complete check, which will come as part of the parallelizability checker.

**Comment by magnatelee** [comment]
@manopapad this issue should be properly fixed as of b29f93d6.
